lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -92,3 +105,377 @@ def gen_month_range(prefix, index, start, end):
         else:
             queryString += f"{prefix + index + monthValue},"
     return queryString
+
+
+def convert_date(value: str):
+    return parser.parse(value)
+
+
+def convert_username(value: str):
+    return Auth.validate_user(value)
+
+
+def convert_json(value: dict):
+    if type(value) is not dict:
+        raise TypeError(f"Unexpected type {type(value).__name__}")
+    return value
+
+
+def convert_string(value: str):
+    return str(value)
+
+
+class ParamType(Enum):
+    """
+    Define the possible JSON query parameter keys, and their type
+
+    The common code can perform conversions on the required parameters with
+    known types.
+    """
+
+    DATE = ("Date", convert_date)
+    USER = ("User", convert_username)
+    JSON = ("Json", convert_json)
+    STRING = ("String", convert_string)
+
+    def __init__(self, name: AnyStr, convert: Callable[[AnyStr], Any]):
+        """
+        ParamType Enum initializer: this uses a mixed-case name string in
+        addition to the conversion method simply because with only the
+        Callable value I ran into naming issues.
+
+        TODO: I'll experiment a bit more with other ways to avoid the
+        problems I encountered, as I'm not actually using this. Perhaps
+        using auto() for the first parameter would work?
+        """
+        self.friendly = name
+        self.convert = convert
+
+    def __str__(self):
+        return self.name
+
+    def convert(self, value):
+        """
+        convert Convert a string value to the expected type
+
+        Args:
+            value: The value of the JSON key
+
+        Returns:
+            Appropriate translated value
+
+        Raises:
+            Any conversion errors
+        """
+        return self.convert(value)
+
+
+class Parameter:
+    """
+    Define the attributes of a parameter using the ParamType ENUM
+
+    Note that a parameter that's "required" must also be non-empty.
+    """
+
+    def __init__(self, name: AnyStr, type: ParamType, required: bool = False):
+        """
+        __init__ Initialize a Parameter object describing a JSON parameter
+        with its type and attributes.
+
+        Args:
+            name: Parameter name
+            type: Parameter type
+            required: whether the parameter is required (default to False)
+        """
+        self.name = name
+        self.type = type
+        self.required = required
+
+    def __str__(self):
+        return f"Parameter<{self.name}:{self.type}>"
+
+
+class Schema:
+    """
+    Define the client input schema for a server query that's based on
+    Elasticsearch.
+    """
+
+    def __init__(self, *parameters: List[Parameter]):
+        """
+        __init__ Specify an interface schema as a list of Parameter objects.
+
+        Args:
+            parameters: a list of Parameter objects
+        """
+        self.parameters = [*parameters]
```

Comment:
Actually I'm creating a list from a tuple, which was giving me troubles in some context I no longer completely recall.

Reply:
That makes reasonable sense, except that the type hint for `parameters` is `List` -- are `List` and `Tuple` compatible, or do the invocations now pass lists instead of tuples?

In any case, if you intend to keep this, I would propose using the `list(parameters)` idiom rather than `[*parameters]` -- I think it's most easily apprehended by the reader and less likely to be inadvertently damaged by editors.